File: R/prior_calculation.R
library(flowClust)

# Define the prior function
prior <- function(x, kappa, Nt = NULL, addCluster = NULL) {
  if (is.null(Nt)) {
    Nt <- nrow(x@z)
  }
  p <- ncol(x@mu)
  K <- x@K
  nu0 <- Ng <- x@w * Nt
  if (all((nu0 * kappa - p - 1) > 0)) {
    Lambda0 <- x@sigma
    for (i in 1:K) {
      Lambda0[i, , ] <- Lambda0[i, , ] * (kappa * nu0[i] - p - 1)
    }
  } else {
    stop("Can't proceed. Prior nu0 is negative for cluster(s) ", 
         paste(which((nu0 - p - 1) > 0), collapse = ","), 
         "\n(p-1) = ", p - 1, ": Try increasing kappa")
  }
  Omega0 <- array(0, c(K, p, p))
  for (i in 1:K) {
    Omega0[i, , ] <- diag(1, p)
    if (p == 1) {
      dS <- x@sigma[i, , ]
      dO <- Omega0[i, , ]
    } else {
      dS <- det(x@sigma[i, , ])
      dO <- det(Omega0[i, , ])
    }
    k <- (dO/dS)^(1/p)
    Omega0[i, , ] <- Omega0[i, , ] * k
    Omega0[i, , ] <- solve(Omega0[i, , ] * Ng[i] * kappa)
  }
  nu0 <- nu0 * kappa
  Mu0 <- x@mu
  lambda <- x@lambda
  w0 <- x@w * Nt
  if (!is.null(addCluster)) {
    for (i in (K + 1):(K + addCluster)) {
      S <- stats::cov(Mu0)
      Lam <- array(0, c(K + 1, p, p))
      om <- array(0, c(K + 1, p, p))
      Mu0 <- rbind(Mu0, colMeans(Mu0))
      for (i in 1:K) {
        om[i, , ] <- Omega0[i, , ]
        Lam[i, , ] <- Lambda0[i, , ]
      }
      om[K + 1, , ] <- diag(1, p)
      Lam[K + 1, , ] <- diag(1, p)
      diag(Lam[K + 1, , ]) <- diag(S)
      diag(om[K + 1, , ]) <- diag(S)
      if (p == 1) {
        dS <- Lam[K + 1, , ]
        dO <- om[K + 1, , ]
      } else {
        dS <- det(Lam[K + 1, , ])
        dO <- det(om[K + 1, , ])
      }
      k <- (dO/dS)^(1/p)
      om[K + 1, , ] <- om[K + 1, , ] * k
      Omega0 <- om
      Lambda0 <- Lam
      nu0 <- c(nu0, p + 2)
      w0 <- c(w0, 1)
      K <- K + 1
    }
  }
  prior <- list(Mu0 = Mu0, Lambda0 = Lambda0, Omega0 = Omega0, 
                w0 = w0, nu0 = nu0, nu = x@nu, lambda = lambda, K = K)
  class(prior) <- "flowClustPrior"
  attr(prior, "lambda") <- lambda
  
  return(prior)
}

File: Python/prior_calculation.py
import numpy as np

def prior(x, kappa, Nt=None, addCluster=None):
    if Nt is None:
        Nt = x['z'].shape[0]
    p = x['mu'].shape[1]
    K = x['K']
    nu0 = Ng = x['w'] * Nt
    if np.all((nu0 * kappa - p - 1) > 0):
        Lambda0 = x['sigma'].copy()
        for i in range(K):
            Lambda0[i] *= (kappa * nu0[i] - p - 1)
    else:
        raise ValueError("Can't proceed. Prior nu0 is negative for cluster(s). Try increasing kappa.")
    
    Omega0 = np.zeros((K, p, p))
    for i in range(K):
        Omega0[i] = np.eye(p)
        if p == 1:
            dS = x['sigma'][i]
            dO = Omega0[i]
        else:
            dS = np.linalg.det(x['sigma'][i])
            dO = np.linalg.det(Omega0[i])
        k = (dO / dS) ** (1 / p)
        Omega0[i] *= k
        Omega0[i] = np.linalg.inv(Omega0[i] * Ng[i] * kappa)
    
    nu0 *= kappa
    Mu0 = x['mu']
    lambda_ = x['lambda']
    w0 = x['w'] * Nt
    
    if addCluster is not None:
        for i in range(K, K + addCluster):
            S = np.cov(Mu0, rowvar=False)
            Lam = np.zeros((K + 1, p, p))
            om = np.zeros((K + 1, p, p))
            Mu0 = np.vstack([Mu0, Mu0.mean(axis=0)])
            for j in range(K):
                om[j] = Omega0[j]
                Lam[j] = Lambda0[j]
            om[K] = np.eye(p)
            Lam[K] = np.eye(p)
            np.fill_diagonal(Lam[K], np.diag(S))
            np.fill_diagonal(om[K], np.diag(S))
            if p == 1:
                dS = Lam[K]
                dO = om[K]
            else:
                dS = np.linalg.det(Lam[K])
                dO = np.linalg.det(om[K])
            k = (dO / dS) ** (1 / p)
            om[K] *= k
            Omega0 = om
            Lambda0 = Lam
            nu0 = np.append(nu0, p + 2)
            w0 = np.append(w0, 1)
            K += 1
    
    prior = {
        'Mu0': Mu0,
        'Lambda0': Lambda0,
        'Omega0': Omega0,
        'w0': w0,
        'nu0': nu0,
        'nu': x['nu'],
        'lambda': lambda_,
        'K': K
    }
    prior['lambda'] = lambda_
    
    return prior

# Example usage (ensure you have the data structure as per your requirement)
# x = {
#     'z': np.array(...),
#     'mu': np.array(...),
#     'sigma': np.array(...),
#     'w': np.array(...),
#     'K': ...,
#     'nu': ...,
#     'lambda': ...
# }
# kappa = ...
# result = prior(x, kappa)
# print(result)
